haystack_experimental/util/utils.py

```diff
@@ -0,0 +1,76 @@
+# SPDX-FileCopyrightText: 2022-present deepset GmbH <info@deepset.ai>
+#
+# SPDX-License-Identifier: Apache-2.0
+
+from typing import Union, List
+
+
+def merge_dicts(d1, d2):
+    """
+    Merge two dictionaries. If a key exists in both dictionaries, the values are merged according to the rules:
+
+    - If both values are dictionaries, merge them recursively
+    - If both values are lists, concatenate them
+    - If the types are different, store both values in a list
+
+    :param d1: Dictionary 1
+    :param d2: Dictionary 2
+    :returns:
+        a single dictionary with the merged values from d1 and d2
+    """
+
+    merged = {}
+
+    def is_primitive(value):
+        return isinstance(value, (int, float, bool, str, type(None)))
+
+    for key in d1.keys():
+        if key in d2:
+            # both are dictionaries
+            if isinstance(d1[key], dict) and isinstance(d2[key], dict):
+                merged[key] = merge_dicts(d1[key], d2[key])
+            # both are lists
+            elif isinstance(d1[key], list) and isinstance(d2[key], list):
+                merged[key] = d1[key] + d2[key]
+            # both are of primitive types
+            elif is_primitive(d1[key]) and is_primitive(d2[key]):
+                merged[key] = [d1[key], d2[key]]
+        else: # key only in d1
+            merged[key] = d1[key]
+
+    # add remaining keys from d2
+    for key in d2.keys():
+        if key not in merged:
+            merged[key] = d2[key]
+
+    return merged
+
+
+def expand_page_range(page_range: Union[ List[Union[int,str]], List[int] ] ):
```

Comment:
Thanks for adding this! One thing that would be nice is if we also support an example like 
```
['1-3', '5', '8', '10-12']
```
where the whole input is a list of strings. Currently when trying this example I get an error at this line
```python
start, end = page.split("-")
```

Reply:
so you are suggesting, essentially, everything as a string

Reply:
Yeah support:
* everything as a string
* everything as a number (when no ranges are specified)
* and I guess the mixture of both